include/swift/SIL/OwnershipLiveness.h

```diff
@@ -0,0 +1,266 @@
+//===--- OwnershipLiveness.h ---------------------------------*- C++ -*----===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+///
+/// Terminology:
+///
+/// Linear lifetime: All paths through the value's definition to the
+/// function exit pass through exactly one lifetime-ending operation.
+///
+/// Complete OSSA: All owned values and borrow introducers have linear lifetime.
+///
+///
+/// Linear SSA liveness:
+///
+/// - Liveness for a single "defining" owned value or borrow introducing value
+///   assuming the value's lifetime is already linear.
+///
+/// - The definition dominates all use points (phis do not extend the lifetime)
+///
+/// - Only lifetime-ending operations generate liveness
+///
+/// - Oblivious to pointer escapes within the lifetime
+///
+///
+/// Interior SSA liveness:
+///
+/// - Liveness for a single "defining" value with any ownership.
+///
+/// - The definition dominates all use points (phis do not extend the lifetime)
+///
+/// - Does not assume the current lifetime is linear. Transitively follows
```

Comment:
```
/// Interior liveness example:
/// 
///     %struct = struct ...
///     %f = struct_extract %s     // defines a guaranteed value
///     %b = begin_borrow %field
///     %a = ref_element_addr %b
///     _  = address_to_pointer %a
///     end_borrow %b              // the only interior use of %f
///
/// When computing interior liveness for %f, %b is an inner scope. Because inner
/// scopes to be complete, the only relevant use is end_borrow %b. Despite the
/// address_to_pointer instruction, %f does not escape any dependent address.
```
** Transitive liveness is no longer needed with complete ownership lifetimes **
```
/// Transitive liveness example:
/// 
///     %struct = struct ...
///     %f = struct_extract %s     // defines a guaranteed value
///     %b = begin_borrow %field
///     %a = ref_element_addr %b
///     _  = address_to_pointer %a // a transitive use of %f escapes
///
/// When computing transitive liveness for %f, %b is an inner scope. Liveness
/// does not assume that an end_borrow exists. Instead it transitively considers
/// all uses of %b. As a result, %f escapes.
```